### ISSUE TITLE
adjusting uses of `int(?w)` etc in the standard library

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#22814]
PR #22769 adds a warning to deprecate implicit conversion when passing to `int(?w)` formal types. At the same time, it left for future work adjusting all of the functions in the standard library.

For each of the standard library functions using `int(?w)`, `uint(?w)`, `real(?w)`, `imag(?w)`, or `complex(?w)`, we need to use one of these alternative paths:
 * if we wish to preserve the ability to have implicit conversions when calling that function; in which case, we stamp out the various concrete versions
 * or, if we do not view implicit conversions when calling the function as important, in which case, we make them generic

**Comment by bradcray** [comment]
This issue seems like it needs a Chapel 2.0 label for routines that are within the set of modules we're trying to stabilize, is that right?

Did you happen to capture a list of the routines that would need to be considered here?  (should be easy enough for me to recreate, but if you already have it...)

**Comment by mppf** [comment]
> This issue seems like it needs a Chapel 2.0 label for routines that are within the set of modules we're trying to stabilize, is that right?

Not necessarily. I don’t think that addressing this is strictly necessary for the Fall release because I think we can either a) say changes in this area are bug fixes or b) manually stamp out these functions in the future to continue to allow implicit conversions. However, as with many things, it would be nicer to accomplish it before the fall release.

> Did you happen to capture a list of the routines that would need to be considered here?

No, I never created such a list. I'm aware of a several `?w` functions in the AutoMath and Math modules, though. `abs` might be the most commonly used one. Also, ti looks like we have unary operators in ChapelBase that would need similar adjustment (e.g. today we can write `-true` but it's not obvious to me if that makes sense).

**Comment by bradcray** [comment]
> b) manually stamp out these functions in the future to continue to allow implicit conversions.

Ah, I think this is what I was missing:  With the current interpretation of `scalarType(?w)`, we've only narrowed the set of signatures that are supported, so if someone's code works with the current signatures it will continue to work going forward, is that right?  I was thinking that potentially the opposite would happen (vacation brain) which feels harder to justify as a bug fix (in the sense that we've documented a given interface as being supported).

**Comment by mppf** [comment]
> With the current interpretation of `scalarType(?w)`, we've only narrowed the set of signatures that are supported

I think we're agreeing, but the old "stamping out" for `scalarType(?w)` still occurs. All I did was add a deprecation warning when an implicit conversion is used when passing to such a stamped-out formal.

> if someone's code works with the current signatures it will continue to work going forward, is that right? 

That's only true if we manually stamp out such functions in the future. The warning I described would not be generated because it is skipped when the formal is question is in an internal/standard module.

**Comment by bradcray** [comment]
> but the old "stamping out" for scalarType(?w) still occurs.

Oh right.  So trying again:  The current set of signatures is too large (current code may break once we no longer stamp it out), but "the bug" is not that the routines' signatures are wrong, but that the compiler stamps out the signatures rather than doing what we think it should do?  (where it's a bug that we won't bother fixing in the production compiler, but will resolve when dyno owns resolution?)

Did I get it right that time?

**Comment by mppf** [comment]
Right, that's how we'd describe this as a bug fix (option a from https://github.com/chapel-lang/chapel/issues/22814#issuecomment-1665771569 ). There is still option b from that comment which is to manually stamp-out the definitions of such functions in the standard library at some point in the future before the compiler changes behavior and stops stamping them out itself.

**Comment by bradcray** [comment]
Though we're agreeing that there's a path that doesn't require 2.0 changes here, I've put the label on anyway so that as we're taking stock of where we are w.r.t. 2.0 tasks, we don't lose sight of this.  If we do decide to consider it a bug, then I think we should create a bug issue specifically about it.
